File: testbenches/testbench.cs
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Reflection;
using System.Diagnostics;
using BP = TestNamespace.EmptyName1.EmptyName2;

namespace CSharpTestbench {
    class Program {

        private static readonly string DATA_FOLDER_NAME = "TestbenchData";
        private static readonly string THIS_LANG = "cs";
        private static readonly string REF_LANG = "py";

        private static readonly int DEFAULT_INT = 114;
        private static readonly float DEFAULT_FLOAT = 114.514f;
        private static readonly string DEFAULT_STR = "shit";
        private static readonly int DEFAULT_LIST_LEN = 5;

        private static readonly int BENCHMARK_TIMES = 100;

        static void GetStructMsgsName(ref List<Type> msgs) {
            var nsname = typeof(BP.OpCode).Namespace;
            var bpmessage_t = typeof(BP.BpMessage);
            var bpstruct_t = typeof(BP.BpStruct);
            var classes = from ty in Assembly.GetExecutingAssembly().GetTypes()
                          where ty.IsClass && ty.Namespace == nsname && ty.IsSubclassOf(bpstruct_t) && ty.Name != bpstruct_t.Name && ty.Name != bpmessage_t.Name
                          select ty;

            msgs.AddRange(from ty in classes
                          where ty.IsSubclassOf(bpmessage_t)
                          select ty);
            //structs.AddRange(from ty in classes
            //                 where (!ty.IsSubclassOf(bpmessage_t))
            //                 select ty);
        }

        static object GenerateDefaultValue(Type t) {
            // process enum first. reveal its unerlying type first
            Type ot = t;
            if (ot.IsEnum) {
                t = t.GetEnumUnderlyingType();
            }

            object instance;
            if (t == typeof(byte)) instance = (byte)DEFAULT_INT;
            else if (t == typeof(sbyte)) instance = (sbyte)DEFAULT_INT;
            else if (t == typeof(ushort)) instance = (ushort)DEFAULT_INT;
            else if (t == typeof(short)) instance = (short)DEFAULT_INT;
            else if (t == typeof(uint)) instance = (uint)DEFAULT_INT;
            else if (t == typeof(int)) instance = DEFAULT_INT;
            else if (t == typeof(ulong)) instance = (ulong)DEFAULT_INT;
            else if (t == typeof(long)) instance = (long)DEFAULT_INT;
            else if (t == typeof(float)) instance = DEFAULT_FLOAT;
            else if (t == typeof(double)) instance = (double)DEFAULT_FLOAT;
            else if (t == typeof(string)) instance = DEFAULT_STR;
            else {
                // this is a complex structure
                instance = Activator.CreateInstance(t);
                AssignVariablesValue(instance);
            }

            // process enum finally
            if (ot.IsEnum) {
                return Enum.ToObject(ot, instance);
            } else {
                return instance;
            }
        }
        static void AssignVariablesValue(object instance) {
            Type instance_t = instance.GetType();

            foreach (var member in instance_t.GetFields(BindingFlags.Public | BindingFlags.Instance)) {
                Type member_t = member.FieldType;

                if (member_t.IsGenericType) {
                    // dynamic array
                    var item_t = member_t.GetGenericArguments()[0];

                    member_t.GetMethod("Clear").Invoke(member.GetValue(instance), null);
                    for (int i = 0; i < DEFAULT_LIST_LEN; ++i) {
                        member_t.GetMethod("Add").Invoke(member.GetValue(instance), new[] { GenerateDefaultValue(item_t) });
                    }
                } else if (member_t.IsArray) {
                    // static array
                    var item_t = member_t.GetElementType();
                    var absarray = (member.GetValue(instance) as Array);

                    for (int i = 0; i < absarray.Length; ++i) {
                        absarray.SetValue(GenerateDefaultValue(item_t), i);
                    }
                } else {
                    // single
                    member.SetValue(instance, GenerateDefaultValue(member_t));
                }

            }
        }

        static bool CompareInstanceMember(Type ty, object member1, object member2) {
            // process enum first
            if (ty.IsEnum) {
                ty = ty.GetEnumUnderlyingType();
            }

            if (ty == typeof(byte)) return (byte)member1 == (byte)member2;
            else if (ty == typeof(sbyte)) return (sbyte)member1 == (sbyte)member2;
            else if (ty == typeof(ushort)) return (ushort)member1 == (ushort)member2;
            else if (ty == typeof(short)) return (short)member1 == (short)member2;
            else if (ty == typeof(uint)) return (uint)member1 == (uint)member2;
            else if (ty == typeof(int)) return (int)member1 == (int)member2;
            else if (ty == typeof(ulong)) return (ulong)member1 == (ulong)member2;
            else if (ty == typeof(long)) return (long)member1 == (long)member2;
            else if (ty == typeof(float)) return (float)member1 == (float)member2;
            else if (ty == typeof(double)) return (double)member1 == (double)member2;
            else if (ty == typeof(string)) return (string)member1 == (string)member2;
            else {
                // this is a complex structure
                return CompareInstance(ty, member1, member2);
            }
        }
        static bool CompareInstance(Type ty, object instance1, object instance2) {
            foreach (var member in ty.GetFields(BindingFlags.Public | BindingFlags.Instance)) {
                // pre check
                object instmem1 = member.GetValue(instance1);
                object instmem2 = member.GetValue(instance2);

                if (instance1.GetType() != instance2.GetType()) return false;

                // data check
                Type member_t = member.FieldType;
                if (member_t.IsGenericType) {
                    // dynamic array
                    var item_t = member_t.GetGenericArguments()[0];

                    // check List<T> length
                    int len1 = (int)member_t.InvokeMember("Count", BindingFlags.Public | BindingFlags.Instance | BindingFlags.GetProperty, null, instmem1, null);
                    int len2 = (int)member_t.InvokeMember("Count", BindingFlags.Public | BindingFlags.Instance | BindingFlags.GetProperty, null, instmem2, null);
                    if (len1 != len2) {
                        return false;
                    }

                    // check every item
                    for (int i = 0; i < len1; ++i) {
                        if (!CompareInstanceMember(item_t,
                            member_t.InvokeMember("Item", BindingFlags.Public | BindingFlags.Instance | BindingFlags.GetProperty, null, instmem1, new object[] { i }),
                            member_t.InvokeMember("Item", BindingFlags.Public | BindingFlags.Instance | BindingFlags.GetProperty, null, instmem2, new object[] { i })
                            )) {
                            return false;
                        }
                    }
                } else if (member_t.IsArray) {
                    // static array
                    var item_t = member_t.GetElementType();

                    var absarray1 = (instmem1 as Array);
                    var absarray2 = (instmem2 as Array);

                    // check length
                    int len1 = absarray1.Length;
                    int len2 = absarray2.Length;
                    if (len1 != len2) {
                        return false;
                    }

                    // check every item
                    for (int i = 0; i < len1; ++i) {
                        if (!CompareInstanceMember(item_t,
                            absarray1.GetValue(i),
                            absarray2.GetValue(i)
                            )) {
                            return false;
                        }
                    }
                } else {
                    // single
                    if (!CompareInstanceMember(member_t, instmem1, instmem2)) {
                        return false;
                    }
                }

            }

            return true;
        }

        static void BenchmarkTest(List<Type> msgs) {
            MemoryStream ms = new MemoryStream();
            ms.Capacity = 1024;

            foreach (var msg in msgs) {
                // create example first
                var instance = (BP.BpMessage)Activator.CreateInstance(msg);
                AssignVariablesValue(instance);
                Console.WriteLine($"Benchmarking {msg.Name}...");

                // create some variables for benchmark
                var watch = new Stopwatch();
                var br = new BinaryReader(ms);
                var bw = new BinaryWriter(ms);
                // pre-run to remove any possible spike
                instance.Serialize(bw);
                ms.Seek(0, SeekOrigin.Begin);
                instance.Deserialize(br);
                ms.SetLength(0);
                BP.BPHelper.UniformSerialize(instance, bw);
                ms.Seek(0, SeekOrigin.Begin);
                BP.BPHelper.UniformDeserialize(br);
                ms.SetLength(0);
                // pure benchmark
                for (int i = 0; i < BENCHMARK_TIMES; ++i) {
                    watch.Start();
                    instance.Serialize(bw);
                    watch.Stop();
                    ms.Seek(0, SeekOrigin.Begin);
                }
                var pureser_msec = watch.Elapsed.TotalMilliseconds / BENCHMARK_TIMES;
                watch.Reset();
                for (int i = 0; i < BENCHMARK_TIMES; ++i) {
                    watch.Start();
                    instance.Deserialize(br);
                    watch.Stop();
                    ms.Seek(0, SeekOrigin.Begin);
                }
                var puredeser_msec = watch.Elapsed.TotalMilliseconds / BENCHMARK_TIMES;
                watch.Reset();
                ms.SetLength(0);
                // non-pure benchmark
                for (int i = 0; i < BENCHMARK_TIMES; ++i) {
                    watch.Start();
                    BP.BPHelper.UniformSerialize(instance, bw);
                    watch.Stop();
                    ms.Seek(0, SeekOrigin.Begin);
                }
                var ser_msec = watch.Elapsed.TotalMilliseconds / BENCHMARK_TIMES;
                watch.Reset();
                for (int i = 0; i < BENCHMARK_TIMES; ++i) {
                    watch.Start();
                    BP.BPHelper.UniformDeserialize(br);
                    watch.Stop();
                    ms.Seek(0, SeekOrigin.Begin);
                }
                var deser_msec = watch.Elapsed.TotalMilliseconds / BENCHMARK_TIMES;
                watch.Reset();
                ms.SetLength(0);

                // output result
                Console.WriteLine($"Serialize (uniform/spec) {ser_msec * 1e3:f7}/{pureser_msec * 1e3:f7} usec. Deserialize (uniform/spec) {deser_msec * 1e3:f7}/{puredeser_msec * 1e3:f7} usec.");
            }
        }

        static bool[] g_EndianSwitches = new bool[] { false, true };
        static string GetEndianStr(bool is_bigendian) {
            return is_bigendian ? "BE" : "LE";
        }
        static string GetFilename(string ident_lang, string ident_msg, bool is_bigendian) {
            return System.IO.Path.Combine(DATA_FOLDER_NAME, ident_lang, GetEndianStr(is_bigendian) + "_" + ident_msg + ".bin");
        }
        static void LangInteractionTest(List<Type> msgs) {
            // create a list. write standard msg. and write them to file
            var standard = new Dictionary<string, BP.BpMessage>();
            foreach (var msg in msgs) {
                var instance = (BP.BpMessage)Activator.CreateInstance(msg);
                AssignVariablesValue(instance);
                standard.Add(msg.Name, instance);

                foreach (bool is_bigendian in g_EndianSwitches) {
                    // open switch
                    BP.BPEndianHelper.g_ForceBigEndian = is_bigendian;
                    // write file
                    using (var fs = new FileStream(GetFilename(THIS_LANG, msg.Name, is_bigendian), FileMode.Create, FileAccess.Write, FileShare.None)) {
                        instance.Serialize(new BinaryWriter(fs));
                    }
                    // reset switch
                    BP.BPEndianHelper.g_ForceBigEndian = false;
                }
            }

            // read from file. and compare it with standard
            foreach (var msg in msgs) {
                Console.WriteLine($"Checking {msg.Name} data correction...");

                // read from file
                var instance = (BP.BpMessage)Activator.CreateInstance(msg);
                foreach (bool is_bigendian in g_EndianSwitches) {
                    // open switch
                    BP.BPEndianHelper.g_ForceBigEndian = is_bigendian;
                    // read file
                    using (var fs = new FileStream(GetFilename(REF_LANG, msg.Name, is_bigendian), FileMode.Open, FileAccess.Read, FileShare.None)) {
                        instance.Deserialize(new BinaryReader(fs));
                    }
                    // reset switch
                    BP.BPEndianHelper.g_ForceBigEndian = false;
                }

                // compare data
                if (!CompareInstance(msg, instance, standard[msg.Name])) {
                    Console.WriteLine("Failed on data correction check!");
                }
            }
        }

        static void Main(string[] args) {
            // get necessary variables
            List<Type> msgs = new List<Type>();
            GetStructMsgsName(ref msgs);

            Console.WriteLine("===== Serialize & Deserialize Benchmark =====");
            BenchmarkTest(msgs);

            Console.WriteLine("===== Language Interaction =====");
            LangInteractionTest(msgs);
        }
    }
}

File: testbenches/testbench.cpp
#include "CodeGenTest.hpp"
#include <cstdio>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <array>
#include <sstream>

namespace BP = TestNamespace::EmptyName1::EmptyName2;

namespace CppTestbench {

	const std::string DATA_FOLDER_NAME = "TestbenchData";
	const std::string THIS_LANG = "cpp";
	const std::string REF_LANG = "py";

	constexpr const uint8_t DEFAULT_INT = UINT8_C(114);
	constexpr const float DEFAULT_FLOAT = 114.514f;
	const std::string DEFAULT_STR = "shit";
	constexpr const uint32_t DEFAULT_LIST_LEN = UINT32_C(5);

	constexpr const int BENCHMARK_TIMES = 100;

	static void* GetPayloadPtr(const std::string& instance_t, void* msgcls) {
		auto search = BP::_BPTestbench_StructLayouts.find(instance_t);
		if (search == BP::_BPTestbench_StructLayouts.end()) throw std::runtime_error("invalid struct name!");

		return reinterpret_cast<char*>(msgcls) + search->second.mPayloadOffset;
	}

	static const std::vector<BP::_BPTestbench_VariableProperty>& PickVariableProperty(const std::string& instance_t) {
		auto search = BP::_BPTestbench_StructLayouts.find(instance_t);
		if (search == BP::_BPTestbench_StructLayouts.end()) throw std::runtime_error("invalid struct name!");
		return search->second.mVariableProperties;
	}

	static void Stream2File(std::stringstream& ss, std::filesystem::path filepath) {
		std::ofstream fs;
		fs.open(filepath, std::ios_base::out | std::ios_base::binary);

		ss.seekg(0, std::ios_base::beg);
		ss.seekp(0, std::ios_base::beg);
		auto inbuf = ss.str();
		fs.write(inbuf.c_str(), inbuf.size());

		fs.close();
	}

	static void File2Stream(std::stringstream& ss, std::filesystem::path filepath) {
		std::ifstream fs;
		fs.open(filepath, std::ios_base::in | std::ios_base::binary);

		std::string obuf;
		fs.seekg(0, std::ios_base::end);
		obuf.resize(static_cast<size_t>(fs.tellg()));
		fs.seekg(0, std::ios_base::beg);
		fs.read(obuf.data(), obuf.size());

		ss.str(obuf);
		ss.seekg(0, std::ios_base::beg);
		ss.seekp(0, std::ios_base::beg);
		ss.clear();
		fs.close();
	}

	static void AssignVariablesValue(const std::string& instance_t, void* instance) {
		auto& entry = PickVariableProperty(instance_t);

		for (const auto& item : entry) {
			// find data ptr according to container type
			void* dataptr = reinterpret_cast<char*>(instance) + item.mOffset;
			size_t datalen = 1;
			switch (item.mContainerType) {
				case BP::_BPTestbench_ContainerType::Single:
					datalen = 1;
					break;
				case BP::_BPTestbench_ContainerType::Tuple:
					datalen = item.mTupleSize;
					break;	// these are plain type.
				case BP::_BPTestbench_ContainerType::List:
					// resize it first
					item.mListResizeFuncPtr(dataptr, static_cast<BP::_BPTestbench_VecSize_t>(DEFAULT_LIST_LEN));
					// calling list function to get addr.
					dataptr = item.mListDataFuncPtr(dataptr);
					datalen = static_cast<size_t>(DEFAULT_LIST_LEN);
					break;
				default: throw std::runtime_error("assert not reach!");
			}

			// assign value according to different type
			for (size_t i = 0; i < datalen; ++i) {
				switch (item.mBasicType) {
					case BP::_BPTestbench_BasicType::u8:
					case BP::_BPTestbench_BasicType::i8:
						*reinterpret_cast<uint8_t*>(dataptr) = static_cast<uint8_t>(DEFAULT_INT);
						break;
					case BP::_BPTestbench_BasicType::u16:
					case BP::_BPTestbench_BasicType::i16:
						*reinterpret_cast<uint16_t*>(dataptr) = static_cast<uint16_t>(DEFAULT_INT);
						break;
					case BP::_BPTestbench_BasicType::u32:
					case BP::_BPTestbench_BasicType::i32:
						*reinterpret_cast<uint32_t*>(dataptr) = static_cast<uint32_t>(DEFAULT_INT);
						break;
					case BP::_BPTestbench_BasicType::u64:
					case BP::_BPTestbench_BasicType::i64:
						*reinterpret_cast<uint64_t*>(dataptr) = static_cast<uint64_t>(DEFAULT_INT);
						break;
					case BP::_BPTestbench_BasicType::f32:
						*reinterpret_cast<float*>(dataptr) = static_cast<float>(DEFAULT_FLOAT);
						break;
					case BP::_BPTestbench_BasicType::f64:
						*reinterpret_cast<double*>(dataptr) = static_cast<double>(DEFAULT_FLOAT);
						break;
					case BP::_BPTestbench_BasicType::bpstr:
						*reinterpret_cast<std::string*>(dataptr) = DEFAULT_STR;
						break;
					case BP::_BPTestbench_BasicType::bpstruct:
						// recursive calling this function
						AssignVariablesValue(item.mComplexType, dataptr);
						break;
					default: throw std::runtime_error("assert not reach!");
				}

				// inc data ptr
				dataptr = reinterpret_cast<char*>(dataptr) + item.mUnitSizeof;
			}
		}
	}

	static bool CompareInstance(const std::string& instance_t, void* instance1, void* instance2) {
		auto& entry = PickVariableProperty(instance_t);

		for (const auto& item : entry) {
			// find data ptr according to container type
			void* dataptr1 = reinterpret_cast<char*>(instance1) + item.mOffset;
			void* dataptr2 = reinterpret_cast<char*>(instance2) + item.mOffset;

			size_t datalen = 1;
			switch (item.mContainerType) {
				case BP::_BPTestbench_ContainerType::Single:
					datalen = 1;
					break;
				case BP::_BPTestbench_ContainerType::Tuple:
					datalen = item.mTupleSize;
					break;	// these are plain type.
				case BP::_BPTestbench_ContainerType::List:
				{
					// check whether their size is equal
					auto listlen1 = item.mListSizeFuncPtr(dataptr1);
					auto listlen2 = item.mListSizeFuncPtr(dataptr2);
					if (listlen1 != listlen2) {
						return false;
					}

					// now, we can safely set data ptr
					datalen = listlen1;
					dataptr1 = item.mListDataFuncPtr(dataptr1);
					dataptr2 = item.mListDataFuncPtr(dataptr2);

					break;
				}
				default: throw std::runtime_error("assert not reach!");
			}

			// compare value according to different type
			for (size_t i = 0; i < datalen; ++i) {
				switch (item.mBasicType) {
					case BP::_BPTestbench_BasicType::u8:
					case BP::_BPTestbench_BasicType::i8:
						if (*reinterpret_cast<uint8_t*>(dataptr1) != *reinterpret_cast<uint8_t*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::u16:
					case BP::_BPTestbench_BasicType::i16:
						if (*reinterpret_cast<uint16_t*>(dataptr1) != *reinterpret_cast<uint16_t*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::u32:
					case BP::_BPTestbench_BasicType::i32:
						if (*reinterpret_cast<uint32_t*>(dataptr1) != *reinterpret_cast<uint32_t*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::u64:
					case BP::_BPTestbench_BasicType::i64:
						if (*reinterpret_cast<uint64_t*>(dataptr1) != *reinterpret_cast<uint64_t*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::f32:
						if (*reinterpret_cast<float*>(dataptr1) != *reinterpret_cast<float*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::f64:
						if (*reinterpret_cast<double*>(dataptr1) != *reinterpret_cast<double*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::bpstr:
						if (*reinterpret_cast<std::string*>(dataptr1) != *reinterpret_cast<std::string*>(dataptr2))
							return false;
						break;
					case BP::_BPTestbench_BasicType::bpstruct:
						// recursive calling this function
						if (!CompareInstance(item.mComplexType, dataptr1, dataptr2))
							return false;
						break;
					default: throw std::runtime_error("assert not reach!");
				}

				// inc data ptr
				dataptr1 = reinterpret_cast<char*>(dataptr1) + item.mUnitSizeof;
				dataptr2 = reinterpret_cast<char*>(dataptr2) + item.mUnitSizeof;
			}

		}

		return true;
	}

	static void BenchmarkTest() {
		std::stringstream buffer;

		for (uint32_t id = 0; id < BP::_BPTestbench_MessageList.size(); ++id) {
			// output info
			printf("Benchmarking %s...\n", BP::_BPTestbench_MessageList[id].c_str());

			// create exmaple instance
			BP::BpMessage* instance = BP::BPHelper::MessageFactory(static_cast<BP::OpCode>(id));
			void* payload = GetPayloadPtr(BP::_BPTestbench_MessageList[id], instance);
			AssignVariablesValue(BP::_BPTestbench_MessageList[id], payload);

			std::chrono::steady_clock::time_point time_start, time_end;

			// pure benchmark
			time_start = std::chrono::steady_clock::now();
			for (int i = 0; i < BENCHMARK_TIMES; ++i) {
				instance->Serialize(buffer);
				buffer.seekp(0, std::ios_base::beg);
			}
			time_end = std::chrono::steady_clock::now();
			double pure_ser_usec = std::chrono::duration<double, std::micro>(time_end - time_start).count() / BENCHMARK_TIMES;

			time_start = std::chrono::steady_clock::now();
			for (int i = 0; i < BENCHMARK_TIMES; ++i) {
				instance->Deserialize(buffer);
				buffer.seekg(0, std::ios_base::beg);
			}
			time_end = std::chrono::steady_clock::now();
			double pure_deser_usec = std::chrono::duration<double, std::micro>(time_end - time_start).count() / BENCHMARK_TIMES;

			// restore instance and clear buffer
			AssignVariablesValue(BP::_BPTestbench_MessageList[id], payload);
			buffer.str("");
			buffer.clear();
			// uniform benchmark
			time_start = std::chrono::steady_clock::now();
			for (int i = 0; i < BENCHMARK_TIMES; ++i) {
				BP::BPHelper::UniformSerialize(buffer, instance);
				buffer.seekp(0, std::ios_base::beg);
			}
			time_end = std::chrono::steady_clock::now();
			double ser_usec = std::chrono::duration<double, std::micro>(time_end - time_start).count() / BENCHMARK_TIMES;

			std::vector<BP::BpMessage*> temp;
			temp.resize(BENCHMARK_TIMES);
			time_start = std::chrono::steady_clock::now();
			for (int i = 0; i < BENCHMARK_TIMES; ++i) {
				temp[i] = BP::BPHelper::UniformDeserialize(buffer);		// it will produce useless instance. free it after benchmark
				buffer.seekg(0, std::ios_base::beg);
			}
			time_end = std::chrono::steady_clock::now();
			for (const auto& ptr : temp) {
				delete ptr;
			}
			double deser_usec = std::chrono::duration<double, std::micro>(time_end - time_start).count() / BENCHMARK_TIMES;

			// output result
			printf("Serialize (uniform/spec) %.7f/%.7f usec. Deserialize (uniform/spec) %.7f/%.7f usec.\n",
				ser_usec, pure_ser_usec, deser_usec, pure_deser_usec
			);

			// clear
			delete instance;
			buffer.str("");
			buffer.clear();
		}
	}

	static std::array<bool, 2> g_ForceBigEndianTuple{ true, false };
	static void LangInteractionTest() {
		std::filesystem::path wd(DATA_FOLDER_NAME);

		// create a list. write standard msg. and write them to file
		std::vector<BP::BpMessage*> standard;
		for (uint32_t id = 0; id < BP::_BPTestbench_MessageList.size(); ++id) {
			auto& msgname = BP::_BPTestbench_MessageList[id];

			BP::BpMessage* instance = BP::BPHelper::MessageFactory(static_cast<BP::OpCode>(id));
			void* payload = GetPayloadPtr(msgname, instance);
			AssignVariablesValue(msgname, payload);
			
			// add into list
			standard.emplace_back(instance);

			// write into file with 2 different endian
			for (auto& is_bigendian : g_ForceBigEndianTuple) {
				// set endian
				BP::BPHelper::ByteSwap::g_ForceBigEndian = is_bigendian;

				// write file
				std::ofstream fs;
				fs.open(wd / THIS_LANG / ((is_bigendian ? "BE_" : "LE_") + msgname + ".bin"), std::ios_base::out | std::ios_base::binary);
				bool hr = instance->Serialize(fs);
				fs.close();
				if (!hr) {
					throw std::runtime_error("unexpected failed on serialization!");
				}

				// reset endian
				BP::BPHelper::ByteSwap::g_ForceBigEndian = false;
			}
		}

		// read from file. and compare it with standard
		for (uint32_t id = 0; id < BP::_BPTestbench_MessageList.size(); ++id) {
			auto& msgname = BP::_BPTestbench_MessageList[id];

			printf("Checking %s data correction...\n", msgname.c_str());

			// read from file with 2 different endian
			for (auto& is_bigendian : g_ForceBigEndianTuple) {
				// set endian
				BP::BPHelper::ByteSwap::g_ForceBigEndian = is_bigendian;

				// read from file
				BP::BpMessage* instance = BP::BPHelper::MessageFactory(static_cast<BP::OpCode>(id));

				std::ifstream fs;
				fs.open(wd / REF_LANG / ((is_bigendian ? "BE_" : "LE_") + msgname + ".bin"), std::ios_base::in | std::ios_base::binary);
				bool hr = instance->Deserialize(fs);
				fs.close();

				// compare data
				if (!hr || !CompareInstance(msgname, GetPayloadPtr(msgname, instance), GetPayloadPtr(msgname, standard[id]))) {
					printf("Failed on data correction check!\n");
				}

				// reset endian
				BP::BPHelper::ByteSwap::g_ForceBigEndian = false;

				// clear buf
				delete instance;
			}
		}

		// free standard
		for (const auto& ptr : standard) {
			delete ptr;
		}
	}

}


int main(int argc, char* args[]) {

	puts("===== Serialize & Deserialize Benchmark =====\n");
	CppTestbench::BenchmarkTest();

	puts("===== Language Interaction =====\n");
	CppTestbench::LangInteractionTest();

	return 0;
}
